torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -228,11 +229,48 @@ struct DeviceTypePropagationPass {
     return changed;
   }
 
-  /*
-  bool processLoop(Node* n) {
+  bool applyTensorProps(
+      const at::ArrayRef<Value*>& src,
+      const at::ArrayRef<Value*>& dst) {
+    TORCH_INTERNAL_ASSERT(src.size() == dst.size());
+    bool changed = false;
+    for (int i = 0; i < dst.size(); i++) {
+      auto src_type = src[i]->type()->cast<TensorType>();
+      changed |= copy_tensor_prop(dst[i], *src_type);
+    }
+    return changed;
+  }
+
+  void processLoop(Node* node) {
     GRAPH_DEBUG("processLoop");
+    auto blocks = node->blocks();
+    auto loop_inner = blocks.at(0); // Need to check this is correct
+
+    // The first two vars are number of iterations, and loop stop condition
+    auto node_var_in = node->inputs().slice(2);
+    // First var is iter_num
+    auto loop_inner_in = loop_inner->inputs().slice(1);
+    // First var is loop stop bool
+    auto loop_inner_out = loop_inner->outputs().slice(1);
+
+    // Apply the inputs to the inside block
+    applyTensorProps(node_var_in, loop_inner_in);
+
+    int iter = 0;
+    for (; iter < 4; iter++) {
+      processBlock(loop_inner);
+      bool inputs_changed =
+          mergeAndApplyTensorProps(node_var_in, loop_inner_out, loop_inner_in);
+      if (!inputs_changed) {
+        break;
+      }
+    }
+    TORCH_INTERNAL_ASSERT(
+        iter < 4, "Failed to apply tensor props to loop due to changing types");
+
+    // Note that the types of loop_inner_in is not the same as loop_inner_out
+    applyTensorProps(loop_inner_in, node->outputs());
```

Comment:
```
applyTensorProps(loop_inner_in, node->outputs());
```

this is a bit subtle. Are you using  `loop_inner_in` because a loop may not run at all?

Reply:
I am using it because it is the merged types of input and loop_inner_out. The types in those two can be different. 
Look at `test_loop_device_change` with input device of `cpu` for an example.